### PR TITLE
Updated trials card

### PR DESCRIPTION
**Pull request by asinusi**
This addresses the changes mentioned in #7072

The "Rounds Won" amount is now shown on the right side (like in game) to allow the number to grow without overlapping the text.

Any feedback on this is appreciated.

image
image

### CODE REVIEW

File: src/app/inventory/store/d2-item-factory.ts [hunk 1]
```diff
@@ -658,6 +673,9 @@ export function makeItem(
 
   return createdItem;
 }
+export function isTrialsPassage(item: DimItem) {
+  return item.objectives?.length === 3 && item.typeName === 'Trials Passage';
```
Comment (delphiactual) [inline src/app/inventory/store/d2-item-factory.ts:677]: This will not work in other locales... in `de` typeName is `Prüfungen-Pass`

Comment (asinusi) [reply src/app/inventory/store/d2-item-factory.ts]: Good point, I should have read the tooltip. I've pushed another change which goes about it in a different way.

File: src/app/inventory/store/d2-item-factory.ts [hunk 2]
```diff
@@ -1,4 +1,6 @@
+import { D1ManifestDefinitions } from 'app/destiny1/d1-definitions';
```
Comment (nev-r) [inline src/app/inventory/store/d2-item-factory.ts:1]: does d1 still have trials?

Comment (nev-r) [reply src/app/inventory/store/d2-item-factory.ts:1]: wait more importantly this is a d2 item building file. it's impossible for the definitions to be d1 definitions

File: src/app/progress/Objective.tsx [hunk 3]
```diff
@@ -62,10 +64,17 @@ export default function Objective({
   }
 
   const isBoolean = isBooleanObjective(objectiveDef, completionValue);
+  const showAsCounter =
+    isTrialsPassage &&
+    objectiveDef.allowOvercompletion &&
+    objectiveDef.progressDescription === 'Rounds Won';
```
Comment (delphiactual) [inline src/app/progress/Objective.tsx:70]: this will also not work with other locales

Comment (delphiactual) [reply src/app/progress/Objective.tsx:70]: isWinsObjective

Comment (delphiactual) [reply src/app/progress/Objective.tsx:70]: if you need string matching consider a PR to d2ai for everything required here... and making a json file to check against instead.

https://github.com/DestinyItemManager/d2-additional-info

File: src/app/inventory/store/d2-item-factory.ts (hunk 2 above)
Comment (asinusi) [reply src/app/inventory/store/d2-item-factory.ts:1]: Yep you're right. I can change the reference in `ItemDetails.tsx` that uses D1 and check it's a D2 def first like this:

`isTrialsPassage={defs.isDestiny2() && isTrialsPassage(item, defs)}`

Comment (nev-r) [reply src/app/inventory/store/d2-item-factory.ts:1]: ahh, i see now it's exported.
yeah if it's a more general helper function, definitely find a nice place to put it that's not d2 item factory file.

File: src/app/progress/Objective.tsx (hunk 3 above)
Comment (asinusi) [reply src/app/progress/Objective.tsx:70]: Thanks for that. For this one I could get it without using a string comparison as it is different to the other objectives within the item, so now I don't have anymore string comparisons.

File: src/app/inventory/store/objectives.ts [hunk 4]
```diff
@@ -61,3 +62,53 @@ export function isBooleanObjective(
...
+
+// Assumes that the item related to the objective is a trials passage
+export function isFlawlessObjective(
+  objective: DestinyObjectiveProgress,
+  objectiveDef: DestinyObjectiveDefinition
+) {
+  return (
+    objective.completionValue === 1 &&
+    objectiveDef.allowValueChangeWhenCompleted &&
+    !objectiveDef.allowOvercompletion
```
Comment (bhollis) [inline src/app/inventory/store/objectives.ts:101]: Are we sure the flawless objective is the only one that could match this?

File: src/app/inventory/store/objectives.ts [hunk 5]
```diff
@@ -61,3 +62,53 @@ export function isBooleanObjective(
...
+      const objectiveDef = defs.Objective.get(objective.objectiveHash);
+      if (isFlawlessObjective(objective, objectiveDef)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+export function isFlawlessPassage(
```
Comment (bhollis) [inline src/app/inventory/store/objectives.ts:78]: I think this needs documentation - I didn't realize until reading the code that it meant "is the trials passage flawless?"

File: src/app/inventory/store/objectives.ts [hunk 6]
```diff
@@ -61,3 +62,53 @@ export function isBooleanObjective(
...
+  objectives: DestinyObjectiveProgress[] | null,
+  defs: D2ManifestDefinitions
+) {
+  if (objectives && objectives.length === 3) {
+    for (const objective of objectives) {
+      if (isFlawlessObjective(objective, defs.Objective.get(objective.objectiveHash))) {
+        return objective.complete;
+      }
+    }
+    return false;
```
Comment (bhollis) [inline src/app/inventory/store/objectives.ts:88]: No need for this one

File: src/app/inventory/store/d2-item-factory.ts [hunk 7]
```diff
@@ -554,10 +559,24 @@ export function makeItem(
   if (createdItem.objectives) {
     const length = createdItem.objectives.length;
     if (length > 0) {
+      const checkTrialsPassage = isTrialsPassage(createdItem, defs);
```
Comment (bhollis) [inline src/app/inventory/store/d2-item-factory.ts:562]: Move this down to where it's used, we don't need to check it unless `objective.completionValue` is truthy

File: src/app/progress/Objective.tsx [hunk 8]
```diff
@@ -62,10 +64,18 @@ export default function Objective({
   }
 
   const isBoolean = isBooleanObjective(objectiveDef, completionValue);
+  const showAsCounter =
+    isTrialsPassage &&
+    objectiveDef.allowOvercompletion &&
+    !objectiveDef.allowNegativeValue &&
+    !objectiveDef.allowValueChangeWhenCompleted;
```
Comment (bhollis) [inline src/app/progress/Objective.tsx:71]: Isn't this in one of your other helpers?

File: src/app/inventory/store/objectives.ts (hunk 4 above)
Comment (asinusi) [reply src/app/inventory/store/objectives.ts:101]: There is no guarantee so I think the best way would be to create a json file with a list of the hashes I need using [https://github.com/DestinyItemManager/d2-additional-info](url) as @delphiactual had originally suggested. That way I can easily check for the hashes and not worry about the objective definitions. I'll make a PR there and then implement the logic in here once it's been approved.

**Comment by delphiactual** [comment]
#7762

File: src/app/progress/Objective.tsx (hunk 8 above)
Comment (asinusi) [reply src/app/progress/Objective.tsx]: This was a different helper that was only used here but I've put it in `objectives.ts` along with the other helpers.

File: src/app/inventory/store/objectives.ts (hunk 5 above)
Comment (asinusi) [reply src/app/inventory/store/objectives.ts]: I've added a comment to explain it.

File: src/app/inventory/store/objectives.ts (hunk 6 above)
Comment (asinusi) [reply src/app/inventory/store/objectives.ts]: This function has been simplified in the new code.

**Comment by delphiactual** [comment]
I will test this more thoroughly once trials is active, for my approval.

**Review by delphiactual — APPROVED** [review]

**Review by bhollis — APPROVED** [review]

**Comment by delphiactual** [comment]
Awesome work. Thank you.

**Comment by asinusi** [comment]
Thanks for the help along the way :)

**Pull request merged**